From 682baa3b419394dfb654be3923f76c573dd1d67c Mon Sep 17 00:00:00 2001
From: Roger Quadros <rogerq@ti.com>
Date: Fri, 12 Sep 2014 15:19:09 +0300
Subject: [PATCH 1191/1587] HACK: ARM: OMAP2+: pdata-quirks: Add am437x-gp-evm
 NAND/eMMC selection

Some boards e.g. am437x-gp-evm, need a GPIO to enable the
right buffer for either NAND or eMMC operation. This falls out
of scope of both the drivers and hence must be done at the board
driver level.

Add the relevant quirk for am437x-gp-evm. The GPIO will
be configured depending on presence/abscense of gpmc node.

Signed-off-by: Roger Quadros <rogerq@ti.com>
[zou: Original patch taken from
ti-sdk-am335x-evm-08.00.00.00-Linux-x86-Install.bin]
Signed-off-by: Cao Zou <cao.zou@windriver.com>
---
 arch/arm/mach-omap2/pdata-quirks.c |   24 ++++++++++++++++++++++++
 1 files changed, 24 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/pdata-quirks.c b/arch/arm/mach-omap2/pdata-quirks.c
index 9dedf23..fc593c6 100644
--- a/arch/arm/mach-omap2/pdata-quirks.c
+++ b/arch/arm/mach-omap2/pdata-quirks.c
@@ -279,6 +279,27 @@ static void __init omap5_uevm_legacy_init(void)
 }
 #endif
 
+#ifdef CONFIG_SOC_AM43XX
+#define AM43x_GP_EVM_EMMCNAND_GPIO 121	/* gpio3_25 set SEL_eMMCorNANDn */
+static void __init am437x_gp_evm_legacy_init(void)
+{
+	struct device_node *node;
+	int ret, flags;
+
+	flags = GPIOF_OUT_INIT_HIGH;	/* Enable eMMC */
+	node = of_find_compatible_node(NULL, NULL, "ti,am3352-gpmc");
+
+	if (node && of_device_is_available(node))
+		flags = GPIOF_OUT_INIT_LOW;	/* Enable NAND */
+
+	ret = gpio_request_one(AM43x_GP_EVM_EMMCNAND_GPIO,
+			       flags, "NAND/eMMC switch");
+	if (ret)
+		pr_err("%s: Failed to get gpio %d\n",
+		       __func__, AM43x_GP_EVM_EMMCNAND_GPIO);
+}
+#endif
+
 static struct pcs_pdata pcs_pdata;
 
 void omap_pcs_legacy_init(int irq, void (*rearm)(void))
@@ -419,6 +440,9 @@ static struct pdata_init pdata_quirks[] __initdata = {
 #ifdef CONFIG_SOC_OMAP5
 	{ "ti,omap5-uevm", omap5_uevm_legacy_init, },
 #endif
+#ifdef CONFIG_SOC_AM43XX
+	{ "ti,am437x-gp-evm", am437x_gp_evm_legacy_init, },
+#endif
 	{ /* sentinel */ },
 };
 
-- 
1.7.5.4

